contracts/Rentals.sol

```diff
@@ -313,15 +275,30 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
     }
 
     /// @notice Standard function called by ERC721 contracts whenever a safe transfer occurs.
-    /// @dev The contract only allows safe transfers by itself made by the rent function.
-    /// @param _operator Caller of the safe transfer function.
+    /// Provides an alternative to acceptOffer by letting the asset holder send the asset to the contract
+    /// and accepting the offer at the same time.
+    /// @param _operator Caller of the safeTransfer function.
+    /// @param _from Address of the original holder of the token.
+    /// @param _data Bytes containing offer data.
     function onERC721Received(
         address _operator,
-        address, // _from,
+        address _from,
         uint256, // _tokenId,
-        bytes calldata // _data
-    ) external view override returns (bytes4) {
-        require(_operator == address(this), "Rentals#onERC721Received: ONLY_ACCEPT_TRANSFERS_FROM_THIS_CONTRACT");
+        bytes memory _data
+    ) external override returns (bytes4) {
+        // When calling acceptListing or acceptOffer, this contract will transfer this asset from the holder to itself to initialize a rent.
+        // There is no need to do anything extra if that is the case.
+        if (_operator != address(this)) {
+            Offer memory offer = abi.decode(_data, (Offer));
+
+            require(_msgSender() == offer.contractAddress, "Rentals#onERC721Received: SENDER_CONTRACT_ADDRESS_MISMATCH");
+
+            // Set the lessor of the rental to track that the contract already has it.
+            rentals[offer.contractAddress][offer.tokenId].lessor = _from;
```

Comment:
I would respect the CEI pattern, moving this after all of the requirements inside `_acceptOffer`